### PR TITLE
Provide a nicer way of casting to a concrete Obj type.

### PR DESCRIPTION
**Pull request by ltratt**
There is some inherent ickiness here, because given a `Val`, you always have to call `gc_obj` in order to get a `ThinObj`. From Rust's perspective, the resulting `ThinObj` is a temporary value (which can be quite confusing when you start borrowing things).

Perhaps surprisingly, this gives us an opportunity: we can't put casting on `Val` because if it's unboxed, we have to create a boxed version, which creates a temporary which we can't then borrow from. However, by putting `cast` on `ThinObj` we sidestep the problem entirely: we get a (relatively) nice way of casting (including generating a `VMError` which records what type we expected, and what we actually got) while avoiding introducing any new borrow checking problems.

### CODE REVIEW

File: src/lib/vm/objects.rs [hunk 1]
```diff
@@ -263,6 +263,17 @@ impl ThinObj {
         let thinptr = (o as *const _ as *const u8).sub(size_of::<ThinObj>()) as *const ThinObj;
         Gc::recover(thinptr)
     }
+
+    /// Cast this `ThinObj` to a concrete `Obj` instance.
+    pub fn cast<T: Obj + 'static>(&self) -> Result<&T, VMError> {
```
Comment (jacob-hughes) [inline src/lib/vm/objects.rs]: Is `cast` the right terminology here? Intuitively something like `as_obj` makes more sense to me as casting implies the user will provide the destination type for conversion.

Comment (ltratt) [reply src/lib/vm/objects.rs]: A (too long but more accurate) name would be `cast_if_possible_or_return_vmerror_typeerror`. Dunno if that helps think of a better name or not ;)

File: src/lib/vm/objects.rs [hunk 2]
```diff
@@ -263,6 +263,17 @@ impl ThinObj {
...
+        self.deref()
+            .as_any()
+            .downcast_ref()
```
Comment (jacob-hughes) [inline src/lib/vm/objects.rs]: So, this method should give us a fat pointer to some `T: Obj`?  It wouldn't hurt to add the type parameter `T` to `downcast_ref` for clarity.

Comment (ltratt) [reply src/lib/vm/objects.rs]: Yes, it gives us a fat pointer.

File: src/lib/vm/objects.rs (hunk 1 above)
Comment (jacob-hughes) [reply src/lib/vm/objects.rs]: Why not `as_obj`?

**Comment by jacob-hughes** [comment]
The code here looks fine, but I'm not quite clear on the motivation. The `Deref` trait for `ThinObj` constructs a fat pointer lazily for us which includes the vptr. Why might we want to cast to an `Obj`?

File: src/lib/vm/objects.rs (hunk 2 above)
Comment (ltratt) [reply src/lib/vm/objects.rs]: Sorry, my answer was completely wrong -- I don't know what I was thinking :(

No, this doesn't give us back a fat pointer. It gives us a thin pointer to a concrete `struct` that `impl`s `Obj`.

Comment (jacob-hughes) [reply src/lib/vm/objects.rs]: Ah okay. This makes more sense now!

File: src/lib/vm/objects.rs (hunk 1 above)
Comment (vext01) [reply src/lib/vm/objects.rs]: I'm OK with just `cast`

**Comment by vext01** [comment]
This looks good as far as I understand the system, but admittedly I also don't really understand the motivation. It's not a criticism of you intention, but rather: I don't really know the system well enough to know what's going on here.

Perhaps tomorrow you can give me (or us) a rundown of how the system hang together?

**Comment by ltratt** [comment]
I will try. I think we can merge this in the interim?

**Comment by vext01** [comment]
It's OK for me, as long as Jake is convinced. He's the GC guy ;)

**Comment by ltratt** [comment]
This isn't a GC PR.

**Comment by vext01** [comment]
> I don't really know the system well enough to know what's going on here

I rest my case 🔨

**Comment by ltratt** [comment]
OK, @jacob-hughes, it's down to you.

**Comment by jacob-hughes** [comment]
LGTM, no edits so I'm guessing you don't need to squash? If so I'll merge

**Comment by ltratt** [comment]
I can't squash 1 commit down any further :p

**Comment by jacob-hughes** [comment]
You might have wanted to update commit message though ;)

**Comment by ltratt** [comment]
That is true. But not in this case. Feel free to merge as-is!

**Pull request merged**